src/output.cc

```diff
@@ -72,6 +73,11 @@ void EmuPlayer::frame()
     }
     i = min(towrite, (long)(minicnt / p->getrefresh() + 4) & ~3);
     opl->update((short *)pos, i);
+    CDiskopl *dopl = dynamic_cast<CDiskopl *>(opl);
+    if (dopl != NULL) {
+      dopl->update(p);
+    }
+
```

Comment:
I think this is due to a bug in AdPlug.  `CDiskopl::update()` only contains one parameter, but it should contain two in order to correctly override `Copl::update()`.

For this I would suggest adding the second parameter to `CDiskopl::update()` so that this bit of the patch is no longer required.

If it works, please submit a PR to the main AdPlug library for this part.

Thanks for sending in the PR, I hope you don't mind my comments!

Reply:
It's not just that, though, because the param it does take is a `CPlayer`, which is incompatible with what `Copl::update` does. I'm sure there's a nicer way to do things, but I'm not sure what it is.

Reply:
(change the signature of `update` to require all three things, and ignore various subsets of them in different OPLs? That has a broad impact though.)

Reply:
CDiskopl->update is on purpose not overriding the virtual update from the Copl base class, since they are for very different things. They should had different names, but a bit late to change that now.


virtual Copl->update(short *buf, int samples) is only used when the audio chain involves an PCM audio stream:
ALSA/OSS/ESD/libao talks with CPlayer to push time and retrieve delays, and pulls audio-data from the Copl->update(buf,length) as needed.

CDiskopl has its own dance. It wants the I/O data written to Copl, but it also wants the delay information that CPlayer provides which does not fit the model that audio-stream uses.
CDiskopl->update() is the common code that the top-level classes like EmuPlayer in AdPlug/Unix would have contained.

I also agree to hijacking NullOutput is a bit dirty, and a more clean implementation would be to add it as its own Player class "DiskPlayer" that always and only is used when emulator type is Emu_Rawout.

Reply:
Yeah, I'll happily admit that this isn't done "right". I didn't quite follow your class hierarchy and C++ isn't a language I touch very often. I'd be grateful for any amount of surgery if it gets this mergeable.

Reply:
They are a bit messy, and update() name is used multiple locations